File: Creational/Factory/kotlin/Factory.kt
interface Shape {
    fun draw();
}

class ShapeFactory {
    fun getShape(shapeType: String) : Shape =
            when(shapeType) {
                "CIRCLE" -> Circle()
                "RECTANGLE" -> Rectanble()
                "SQUARE" -> Square()
            }
}

class Circle : Shape {
    override fun draw() {
        println("Inside Circle::draw() method");
    }
}

class Rectangle : Shape {
    override fun draw() {
        println("Inside Rectangle::draw() method");
    }
}

class Square : Shape {
    override fun draw() {
        println("Inside Square::draw() method");
    }
}

fun main(args: Array<String>) {
    var shapeFactory = ShapeFactory();
    shapeFactory.getShape("CIRCLE");
    shapeFactory.getShape("RECTANGLE");
    shapeFactory.getShape("SQUARE");
}

File: Creational/Factory/swift/factory.swift
enum ShapeType {
    case circle
    case rectangle
    case square
}

class ShapeFactory {
    func make(shape: ShapeType) -> Shape {
        switch shape {
        case .circle: return Circle.make()
        case .rectangle: return  Rectangle.make()
        case .square: return Square.make()
        }
    }
}

protocol Shape {
    static func make() -> Shape
}

class Circle: Shape {
    static func make() -> Shape {
        return Circle()
    }
}

class Rectangle: Shape {
    static func make() -> Shape {
        return Rectangle()
    }
}

class Square: Shape {
    static func make() -> Shape {
        return Square()
    }
}


var shapeFactory = ShapeFactory()
shapeFactory.make(shape: .circle)
shapeFactory.make(shape: .rectangle)
shapeFactory.make(shape: .square)
